### ISSUE TITLE
Firefox and Chromium "set as default" not including wrapper

### ISSUE THREAD

**Issue by VShell** [NixOS/nixpkgs#4370]
When "set as default browser" is clicked from within Firefox, it places a custom .desktop file in ~/.local/share/applications, and points to this from mimeapps.list in the same directory. This file launches Firefox without the wrapper when it's opened using xdg-open. The result is that if xdg-open is used when no Firefox window exists, it will be opened without environment variables set correctly, and without plugins.

**Comment by benley** [comment]
There's still no obvious way to set the default browser that I can find - anyone know what the status is here?

**Comment by VShell** [comment]
A possibly reasonable option in the case of Firefox would be to disable the default browser check using the `mozilla.cfg` preferences file, and write a custom dialog for selecting the default browser in the wrapper itself, a very rough basis being:

```bash
if check_is_not_default_browser && [ ! -e ~/.config/nix-wrappers/firefox/never-ask ]; then
  yad --title="Default browser" --text="Would you like to set Firefox as your default browser?" --button="gtk-ok:0" --button="gtk-cancel:1" --button="Never:2" --center
  if [ $? -eq 0 ]; then
    xdg-mime default firefox.desktop text/html
    xdg-mime default firefox.desktop x-scheme-handler/http
    xdg-mime default firefox.desktop x-scheme-handler/https
    xdg-mime default firefox.desktop x-scheme-handler/about
  elif [ $? -eq 2 ]; then
    echo 1 > ~/.config/nix-wrappers/firefox/never-ask
  fi
fi
```

According to a google, similar should work for Chrome/Chromium, except passing `--no-first-run --no-default-browser-check` instead of messing with config files.

Opinions?

**Comment by layus** [comment]
Is there no way to trick firefox into thinking its executable name is the wrapper ?

**Comment by vcunat** [comment]
Yes: 0834dbc126 but this needs some testing, and it affects many browsers.

**Comment by layus** [comment]
@vcunat I fixed the wrapper to work with 0834dbc (https://github.com/vcunat/nixpkgs/pull/2), but we get another issue.

When we update firefox, we end up with a different /nix path, which firefox considers as a different browser. If you restart firefox (for example when installing addons) then you get the not-your-default-browser" warning.

Investaigating MOZ_APP_LAUNCHER as solution, but I start to think that https://github.com/NixOS/nixpkgs/issues/4370#issuecomment-168336980 may be the most general solution.

**Comment by KarinaBauer** [comment]
I have the same problem with opera
